server/src/test/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitorTest.java

```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+
+    @Test
+    public void testGetEntityClass() {
+        ProductNodeVisitor visitor = this.buildProductNodeVisitor();
+
+        assertEquals(Product.class, visitor.getEntityClass());
+    }
+
+    @ParameterizedTest(name = "{displayName} {index}: {0}")
+    @MethodSource("productDataProvider")
+    public void testProcessNodeForSkippedEntity(String key, Object base, Object update) {
+        Owner owner = TestUtil.createOwner();
+        String id = TestUtil.randomString("test_id");
+
+        Product existing = this.createPopulatedExistingEntity(id, key, base);
+
+        EntityNode<Product, ProductInfo> node = new ProductNode(owner, id)
+            .setExistingEntity(existing);
+
+        ProductNodeVisitor visitor = this.buildProductNodeVisitor();
+
+        // Ensure initial node state
+        assertFalse(node.visited());
+        assertFalse(node.changed());
+        assertNull(node.getMergedEntity());
+
+        // Visit/process the node
+        visitor.processNode(this.mockNodeProcessor, this.mockNodeMapper, node);
+        visitor.complete();
+
+        // Validate "processed" state
+        assertTrue(node.visited());
+        assertFalse(node.changed());
+        assertNull(node.getMergedEntity());
+    }
+
+    @ParameterizedTest(name = "{displayName} {index}: {0}")
+    @MethodSource("productDataProvider")
+    public void testProcessNodeForUnmodifiedEntity(String key, Object base, Object update) {
+        Owner owner = TestUtil.createOwner();
+        String id = TestUtil.randomString("test_id");
+
+        Product existing = this.createPopulatedExistingEntity(id, key, base);
+        ProductInfo imported = this.createPopulatedImportedEntity(id, key, null);
+
+        EntityNode<Product, ProductInfo> node = new ProductNode(owner, id)
+            .setExistingEntity(existing)
+            .setImportedEntity(imported);
+
+        ProductNodeVisitor visitor = this.buildProductNodeVisitor();
+
+        // Ensure initial node state
+        assertFalse(node.visited());
+        assertFalse(node.changed());
+        assertNull(node.getMergedEntity());
+
+        // Visit/process the node
+        visitor.processNode(this.mockNodeProcessor, this.mockNodeMapper, node);
+        visitor.complete();
+
+        // Validate "processed" state
+        assertTrue(node.visited());
+        assertFalse(node.changed());
+        assertNull(node.getMergedEntity());
+    }
+
+    @ParameterizedTest(name = "{displayName} {index}: {0}")
+    @MethodSource("productDataProvider")
+    public void testProcessNodeForUnchangedEntity(String key, Object base, Object update) {
+        Owner owner = TestUtil.createOwner();
+        String id = TestUtil.randomString("test_id");
+
+        Product existing = this.createPopulatedExistingEntity(id, key, base);
+        ProductInfo imported = this.createPopulatedImportedEntity(id, key, base);
+
+        MergeValidator<ProductInfo> validator = this.validators.get(key);
```

Comment:
Same here, validator not used after retrieving here

Reply:
Fixed